dkg/pedersen/handler.go

```diff
@@ -89,97 +119,160 @@ func (h *Handler) Stream(out mino.Sender, in mino.Receiver) error {
-	responses := []*pedersen.Response{}
-
-mainSwitch:
-	from, msg, err := in.Recv(context.Background())
-	if err != nil {
-		return xerrors.Errorf("failed to receive: %v", err)
-	}
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
 
-	// We expect a Start message or a decrypt request at first, but we might
-	// receive other messages in the meantime, like a Deal.
-	switch msg := msg.(type) {
-
-	case types.Start:
-		err := h.start(msg, deals, responses, from, out, in)
+	for {
+		from, msg, err := in.Recv(ctx)
 		if err != nil {
-			return xerrors.Errorf("failed to start: %v", err)
+			if h.startRes.Done() {
+				// successfully started, terminate cleanly
+				return nil
+			}
+			return xerrors.Errorf("failed to receive: %v", err)
 		}
 
-	case types.Deal:
-		// This is a special case where a DKG started, some nodes received the
-		// start signal and started sending their deals but we have not yet
-		// received our start signal. In this case we collect the Deals while
-		// waiting for the start signal.
-		deals = append(deals, msg)
-		goto mainSwitch
-
-	case types.Response:
-		// This is a special case where a DKG started, some nodes received the
-		// start signal and started sending their deals but we have not yet
-		// received our start signal. In this case we collect the Response while
-		// waiting for the start signal.
-		response := &pedersen.Response{
-			Index: msg.GetIndex(),
-			Response: &vss.Response{
-				SessionID: msg.GetResponse().GetSessionID(),
-				Index:     msg.GetResponse().GetIndex(),
-				Status:    msg.GetResponse().GetStatus(),
-				Signature: msg.GetResponse().GetSignature(),
-			},
-		}
-		responses = append(responses, response)
-		goto mainSwitch
+		dela.Logger.Trace().Msgf("%v received message from %v\n", h.me, from)
+
+		// We expect a Start message or a decrypt request at first, but we might
+		// receive other messages in the meantime, like a Deal.
+		switch msg := msg.(type) {
 
-	case types.DecryptRequest:
-		if !h.startRes.Done() {
-			return xerrors.Errorf("you must first initialize DKG. Did you " +
-				"call setup() first?")
+		case types.Start:
+			dela.Logger.Trace().Msgf("%v received start from %v\n", h.me, from)
+			done := make(chan struct{})
+			err = h.handleStart(out, from, msg, done)
+			if err != nil {
+				return xerrors.Errorf("failed to handle start: %v", err)
+			}
+			go func() {
+				<-done
+				if h.startRes.Done() {
+					cancel()
+				}
+			}()
+
+		case types.Deal:
+			dela.Logger.Trace().Msgf("%v received deal from %v\n", h.me, from)
+
+			h.deals <- dealFrom{
+				&msg,
+				from,
+			}
+
+		case types.Response:
+			dela.Logger.Trace().Msgf("%v received response from %v\n", h.me, from)
+
+			h.responses <- responseFrom{
+				&msg,
+				from,
+			}
+
+		case types.DecryptRequest:
+			dela.Logger.Trace().Msgf("%v received decrypt request from %v\n", h.me, from)
+
+			return h.handleDecrypt(out, msg, from)
+
+		default:
+			dela.Logger.Error().Msgf(
```

Comment:
we usually don't use this kind of "multiline" function call style in our go code. Also, the `\n` is not necessary: 

```go
		dela.Logger.Error().Msgf("%v received an unsupported message %v from %v", 
			h.me, msg, from)
```

(same in other places)